homeassistant/components/tfl/sensor.py

```diff
@@ -0,0 +1,175 @@
+"""Sensor for Transport for London (TfL)."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from operator import itemgetter
+import typing
+from urllib.error import HTTPError, URLError
+
+from tflwrapper import stopPoint
+
+from homeassistant.components.sensor import (  # ENTITY_ID_FORMAT,; PLATFORM_SCHEMA,
+    SensorEntity,
+)
+from homeassistant.const import UnitOfTime
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import TfLConfigEntry
+from .common import call_tfl_api
+from .const import CONF_STOP_POINTS, DOMAIN
+
+ATTR_NEXT_ARRIVALS = "all"
+ATTR_NEXT_THREE_ARRIVALS = "next_3"
+
+RAW_ARRIVAL_LINE_NAME = "lineName"
+RAW_ARRIVAL_DESTINATION_NAME = "destinationName"
+RAW_ARRIVAL_TIME_TO_STATION = "timeToStation"
+
+SCAN_INTERVAL = timedelta(seconds=30)
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: TfLConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the TfL sensor(s)."""
+    stop_point_api = config_entry.runtime_data
+
+    stop_point_ids: list[str] = config_entry.options[CONF_STOP_POINTS]
+
+    unique_id = config_entry.unique_id
+
+    if typing.TYPE_CHECKING:
+        assert unique_id
+
+    try:
+        stop_point_infos = await call_tfl_api(
+            hass, stop_point_api.getByIDs, stop_point_ids, False
```

Comment:
This code has bugs. The `getByIDs` is not good for this use case.

There are the following problems:
1. Some stop point ids are edge cases such that TfL's stop point hierarchy might return a parent stop point id where the requested stop point id is a child.
2. The API can return the stop points in a different order than requested

So you could want to monitor 3 stop points, but the API might return only 2 because it considers one stop point a child… so this code could get complex to correctly process the response.
Given that all this is trying to do is to obtain the `commonName` for each stop point, it is going to be way simpler to instead use the `getStationArrivals` which takes all that complexity away. The tradeoff is needing to make multiple calls but the simplicity is worth it.

Reply:
Fixed in d8120d885603cc583a491248c5763cca7af6d4a5